Add regular expression processing for SD Refine mode.

Lisanjin opened the pull request:
<!-- Put X in the box below to confirm -->

## Checklist:

- [x ] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md). 

I added a feature to the Stable Diffusion image generation function that allows prompt preprocessing using regular expressions. This feature has been integrated into the Refine mode.

This is my first time submitting a PR on GitHub, so please forgive me if there are any mistakes.

Lisanjin committed:
Add regular expression processing for Refine mode.

Cohee1207 commented:
1. Please change the target branch to staging.
2. What exactly is the goal of this change? How would anyone use it? Will you be ready to provide user documentation for this?
3. If you need to parse regex from string, there's a util function in /public/scripts/utils.js:

https://github.com/SillyTavern/SillyTavern/blob/release/public/scripts/utils.js#L1044-L1059

Lisanjin commented:
Thank you for your reply.

When we use the "generate image" feature above the message, the entire message is passed as the generate image prompt to Stable Diffusion. This is obviously an error, as the message may contain conversations, thought processes, etc., which are clearly unrelated to generating images.

![2830949701d200522b06c1239a0cf516](https://github.com/user-attachments/assets/8f6eba3f-fc64-4547-9c57-6b8ec776caa9)

The Image Generation extension settings provide an "Edit prompts before generation" option. When enabled, it allows users to manually edit the message, remove unnecessary information, and send it as a prompt to Stable Diffusion. However, this feature is quite inconvenient. Therefore, I have added an additional feature to the "Edit prompts before generation" option of the Image Generation extension, allowing users to quickly match image generation prompts in the message using regular expressions.

![71f0ed6664ecaaa59a6ccb7218911651](https://github.com/user-attachments/assets/72939dcb-050a-4e8b-b4b5-7c72476449f6)

![481452e2b29f4a7cffd030f83cd0b22d](https://github.com/user-attachments/assets/b1269952-e434-4d8b-b7c5-bbfbe1358225)


Regarding the utils you mentioned, I don't think it's necessary. I referred to this function in the same file: https://github.com/SillyTavern/SillyTavern/blob/release/public/scripts/extensions/stable-diffusion/index.js#L370

Lastly, I have changed the target branch to staging. As for the 
user documentation you mentioned, I’m not entirely sure what’s needed. Could you please clarify what I should do?

Lisanjin commented:
Thank you for your reply.

When we use the "generate image" feature above the message, the entire message is passed as the generate image prompt to Stable Diffusion. This is obviously an error, as the message may contain conversations, thought processes, etc., which are clearly unrelated to generating images.

![2e320916-a60a-450b-8b1b-fe45103df4dd](https://github.com/user-attachments/assets/1baf55d9-36a0-48ab-92a1-636be88d60de)

The Image Generation extension settings provide an "Edit prompts before generation" option. When enabled, it allows users to manually edit the message, remove unnecessary information, and send it as a prompt to Stable Diffusion. However, this feature is quite inconvenient. Therefore, I have added an additional feature to the "Edit prompts before generation" option of the Image Generation extension, allowing users to quickly match image generation prompts in the message using regular expressions.

![f426b289-4363-4122-956c-ed23a4b817da](https://github.com/user-attachments/assets/b0bbec19-6033-4b4e-b3e8-dbf184329230)

![9a0b35d2-38d6-447b-9e4a-4746513be007](https://github.com/user-attachments/assets/b1c67708-edc3-48d1-a609-688df16e256d)


Regarding the utils you mentioned, I don't think it's necessary. I referred to this function in the same file: https://github.com/SillyTavern/SillyTavern/blob/release/public/scripts/extensions/stable-diffusion/index.js#L370

Lastly, I have changed the target branch to staging. As for the
user documentation you mentioned, I’m not entirely sure what’s needed. Could you please clarify what I should do?

Cohee1207 commented:
> This is obviously an error, as the message may contain conversations, thought processes, etc., which are clearly unrelated to generating images.

Sounds like an issue with the prompt. The added instruction for image generation should exclude any possibility to response with conversations, etc. Thinking process from the reasoning feature is trimmed out when you use auto-parse. Have you tried reenforcing the instruction with phrases like "Only provide N", "Ignore previous instructions", etc.?

> However, this feature is quite inconvenient. Therefore, I have added an additional feature to the "Edit prompts before generation" option of the Image Generation extension, allowing users to quickly match image generation prompts in the message using regular expressions.

I think the label and the default don't communicate the intended use for the feature. More over, the provided default value would not work with any of the default prompt. It seems derived from your customized prompts. An average user would be equally confused. My opinion - you should leave this field empty by default.

>  I referred to this function in the same file

This function builds an internal regex that is not meant to be user-editable. When providing user input for regular expressions, it's better to use a well-known syntax that supports flags and forward slashes, similarly to other areas that provide input for regex.

> user documentation you mentioned, I’m not entirely sure what’s needed. Could you please clarify what I should do?

Here's a page for Image Generation extension in the docs repository. Documenting the new option similarly to the others is what would need to be done when the development is complete.

https://github.com/SillyTavern/SillyTavern-Docs/blob/main/extensions/Stable-Diffusion.md

closuretxt commented:
> Sounds like an issue with the prompt. The added instruction for image generation should exclude any possibility to response with conversations, etc. Thinking process from the reasoning feature is trimmed out when you use auto-parse. Have you tried reenforcing the instruction with phrases like "Only provide N", "Ignore previous instructions", etc.?

Some models are a little more straight with their reasoning, such as Gemini and Claude, using tokens to warm up the response.
You can see better results for SD prompts by adding XML tags to reason before finally sending out the tag list, my problem is similar to his, but I use it directly on my SD prompt instead.
As you mentioned, it is a prompt issue, but allowing more freedom towards formatting SD responses could bring better opportunities for everyone to play around with SD prompts.

I am unrelated to this pull, just thought the feature was useful.

coffeegrind123 commented:
Leaving these here since its kinda related, not directly though

https://github.com/SillyTavern/SillyTavern/pull/4167
https://github.com/coffeegrind123/SillyTavern-Tag-Autocompletion

closuretxt commented:
> Leaving these here since its kinda related, not directly though
> 
> #4167 https://github.com/coffeegrind123/SillyTavern-Tag-Autocompletion

Seems like a extremely solid project! Please post them at the Discord https://discord.gg/fBXxgPPEVy and if possible at the reddit https://www.reddit.com/r/SillyTavernAI/
Everyone will love it, thanks for the awesome work.
